# **Laptop Price Prediction**

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sn

from sklearn.model_selection import train_test_split,KFold,StratifiedKFold,cross_val_score,GridSearchCV
from sklearn import preprocessing
from sklearn.preprocessing import StandardScaler

from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor,AdaBoostRegressor,GradientBoostingRegressor
from xgboost import XGBRegressor
from sklearn.neighbors import KNeighborsRegressor
from sklearn.svm import SVR

In [ ]:
df=pd.read_csv('/content/drive/MyDrive/Camerin DSML/laptops.csv')
df.head()

Unnamed: 0                                           img_link  \
0           0  https://rukminim1.flixcart.com/image/312/312/x...   
1           1  https://rukminim1.flixcart.com/image/312/312/x...   
2           2  https://rukminim1.flixcart.com/image/312/312/l...   
3           3  https://rukminim1.flixcart.com/image/312/312/x...   
4           4  https://rukminim1.flixcart.com/image/312/312/x...   

                                       name  price(in Rs.)  \
0             Lenovo Intel Core i5 11th Gen          62990   
1            Lenovo V15 G2 Core i3 11th Gen          37500   
2      ASUS TUF Gaming F15 Core i5 10th Gen          49990   
3  ASUS VivoBook 15 (2022) Core i3 10th Gen          33990   
4                   Lenovo Athlon Dual Core          18990   

                            processor             ram  \
0  Intel Core i5 Processor (11th Gen)  16 GB DDR4 RAM   
1  Intel Core i3 Processor (11th Gen)   8 GB DDR4 RAM   
2  Intel Core i5 Processor (10th Gen)   8 GB DDR4 RAM   
3  Intel Core i3 Processor (10th Gen)   8 GB DDR4 RAM   
4      AMD Athlon Dual Core Processor   4 GB DDR4 RAM   

                                   os              storage  display(in inch)  \
0         Windows 11 Operating System           512 GB SSD              15.6   
1  64 bit Windows 11 Operating System  1 TB HDD|256 GB SSD              15.6   
2         Windows 11 Operating System           512 GB SSD              15.6   
3  64 bit Windows 11 Operating System           512 GB SSD              15.6   
4                DOS Operating System           256 GB SSD              14.0   

   rating  no_of_ratings  no_of_reviews  
0     4.5           14.0            1.0  
1     4.4           53.0            3.0  
2     4.4         4733.0          463.0  
3     4.3        10406.0         1040.0  
4     3.8           18.0            3.0

**Preprocessing**

In [ ]:
df=df.drop(columns=['img_link','Unnamed: 0'])
df.head()

name  price(in Rs.)  \
0             Lenovo Intel Core i5 11th Gen          62990   
1            Lenovo V15 G2 Core i3 11th Gen          37500   
2      ASUS TUF Gaming F15 Core i5 10th Gen          49990   
3  ASUS VivoBook 15 (2022) Core i3 10th Gen          33990   
4                   Lenovo Athlon Dual Core          18990   

                            processor             ram  \
0  Intel Core i5 Processor (11th Gen)  16 GB DDR4 RAM   
1  Intel Core i3 Processor (11th Gen)   8 GB DDR4 RAM   
2  Intel Core i5 Processor (10th Gen)   8 GB DDR4 RAM   
3  Intel Core i3 Processor (10th Gen)   8 GB DDR4 RAM   
4      AMD Athlon Dual Core Processor   4 GB DDR4 RAM   

                                   os              storage  display(in inch)  \
0         Windows 11 Operating System           512 GB SSD              15.6   
1  64 bit Windows 11 Operating System  1 TB HDD|256 GB SSD              15.6   
2         Windows 11 Operating System           512 GB SSD              15.6   
3  64 bit Windows 11 Operating System           512 GB SSD              15.6   
4                DOS Operating System           256 GB SSD              14.0   

   rating  no_of_ratings  no_of_reviews  
0     4.5           14.0            1.0  
1     4.4           53.0            3.0  
2     4.4         4733.0          463.0  
3     4.3        10406.0         1040.0  
4     3.8           18.0            3.0

In [ ]:
df['name']=df['name'].str.upper()
df.head()

name  price(in Rs.)  \
0             LENOVO INTEL CORE I5 11TH GEN          62990   
1            LENOVO V15 G2 CORE I3 11TH GEN          37500   
2      ASUS TUF GAMING F15 CORE I5 10TH GEN          49990   
3  ASUS VIVOBOOK 15 (2022) CORE I3 10TH GEN          33990   
4                   LENOVO ATHLON DUAL CORE          18990   

                            processor             ram  \
0  Intel Core i5 Processor (11th Gen)  16 GB DDR4 RAM   
1  Intel Core i3 Processor (11th Gen)   8 GB DDR4 RAM   
2  Intel Core i5 Processor (10th Gen)   8 GB DDR4 RAM   
3  Intel Core i3 Processor (10th Gen)   8 GB DDR4 RAM   
4      AMD Athlon Dual Core Processor   4 GB DDR4 RAM   

                                   os              storage  display(in inch)  \
0         Windows 11 Operating System           512 GB SSD              15.6   
1  64 bit Windows 11 Operating System  1 TB HDD|256 GB SSD              15.6   
2         Windows 11 Operating System           512 GB SSD              15.6   
3  64 bit Windows 11 Operating System           512 GB SSD              15.6   
4                DOS Operating System           256 GB SSD              14.0   

   rating  no_of_ratings  no_of_reviews  
0     4.5           14.0            1.0  
1     4.4           53.0            3.0  
2     4.4         4733.0          463.0  
3     4.3        10406.0         1040.0  
4     3.8           18.0            3.0

In [ ]:
df['processor']=df['processor'].str.upper()
df.head()

name  price(in Rs.)  \
0             LENOVO INTEL CORE I5 11TH GEN          62990   
1            LENOVO V15 G2 CORE I3 11TH GEN          37500   
2      ASUS TUF GAMING F15 CORE I5 10TH GEN          49990   
3  ASUS VIVOBOOK 15 (2022) CORE I3 10TH GEN          33990   
4                   LENOVO ATHLON DUAL CORE          18990   

                            processor             ram  \
0  INTEL CORE I5 PROCESSOR (11TH GEN)  16 GB DDR4 RAM   
1  INTEL CORE I3 PROCESSOR (11TH GEN)   8 GB DDR4 RAM   
2  INTEL CORE I5 PROCESSOR (10TH GEN)   8 GB DDR4 RAM   
3  INTEL CORE I3 PROCESSOR (10TH GEN)   8 GB DDR4 RAM   
4      AMD ATHLON DUAL CORE PROCESSOR   4 GB DDR4 RAM   

                                   os              storage  display(in inch)  \
0         Windows 11 Operating System           512 GB SSD              15.6   
1  64 bit Windows 11 Operating System  1 TB HDD|256 GB SSD              15.6   
2         Windows 11 Operating System           512 GB SSD              15.6   
3  64 bit Windows 11 Operating System           512 GB SSD              15.6   
4                DOS Operating System           256 GB SSD              14.0   

   rating  no_of_ratings  no_of_reviews  
0     4.5           14.0            1.0  
1     4.4           53.0            3.0  
2     4.4         4733.0          463.0  
3     4.3        10406.0         1040.0  
4     3.8           18.0            3.0

In [ ]:
df['os']=df['os'].str.upper()
df.head()

name  price(in Rs.)  \
0             LENOVO INTEL CORE I5 11TH GEN          62990   
1            LENOVO V15 G2 CORE I3 11TH GEN          37500   
2      ASUS TUF GAMING F15 CORE I5 10TH GEN          49990   
3  ASUS VIVOBOOK 15 (2022) CORE I3 10TH GEN          33990   
4                   LENOVO ATHLON DUAL CORE          18990   

                            processor             ram  \
0  INTEL CORE I5 PROCESSOR (11TH GEN)  16 GB DDR4 RAM   
1  INTEL CORE I3 PROCESSOR (11TH GEN)   8 GB DDR4 RAM   
2  INTEL CORE I5 PROCESSOR (10TH GEN)   8 GB DDR4 RAM   
3  INTEL CORE I3 PROCESSOR (10TH GEN)   8 GB DDR4 RAM   
4      AMD ATHLON DUAL CORE PROCESSOR   4 GB DDR4 RAM   

                                   os              storage  display(in inch)  \
0         WINDOWS 11 OPERATING SYSTEM           512 GB SSD              15.6   
1  64 BIT WINDOWS 11 OPERATING SYSTEM  1 TB HDD|256 GB SSD              15.6   
2         WINDOWS 11 OPERATING SYSTEM           512 GB SSD              15.6   
3  64 BIT WINDOWS 11 OPERATING SYSTEM           512 GB SSD              15.6   
4                DOS OPERATING SYSTEM           256 GB SSD              14.0   

   rating  no_of_ratings  no_of_reviews  
0     4.5           14.0            1.0  
1     4.4           53.0            3.0  
2     4.4         4733.0          463.0  
3     4.3        10406.0         1040.0  
4     3.8           18.0            3.0

In [ ]:
df.isna().sum()

name                  0
price(in Rs.)         0
processor             0
ram                   0
os                    0
storage               0
display(in inch)      0
rating              296
no_of_ratings       296
no_of_reviews       296
dtype: int64

In [ ]:
df.shape

(984, 10)

In [ ]:
df['rating']=df['rating'].fillna((df['rating'].mean()))
df['no_of_ratings']=df['no_of_ratings'].fillna((df['no_of_ratings'].mean()))
df['no_of_reviews']=df['no_of_reviews'].fillna((df['no_of_reviews'].mean()))

In [ ]:
df.isna().sum()

name                0
price(in Rs.)       0
processor           0
ram                 0
os                  0
storage             0
display(in inch)    0
rating              0
no_of_ratings       0
no_of_reviews       0
dtype: int64

In [ ]:
df.dtypes

name                 object
price(in Rs.)         int64
processor            object
ram                  object
os                   object
storage              object
display(in inch)    float64
rating              float64
no_of_ratings       float64
no_of_reviews       float64
dtype: object

In [ ]:
df['name'].unique()

array(['LENOVO INTEL CORE I5 11TH GEN', 'LENOVO V15 G2 CORE I3 11TH GEN',
       'ASUS TUF GAMING F15 CORE I5 10TH GEN',
       'ASUS VIVOBOOK 15 (2022) CORE I3 10TH GEN',
       'LENOVO ATHLON DUAL CORE', 'APPLE 2020 MACBOOK AIR M1',
       'ASUS VIVOBOOK 14 (2021) CELERON DUAL CORE',
       'DELL VOSTRO RYZEN 3 QUAD CORE 5425U',
       'REDMIBOOK PRO CORE I5 11TH GEN',
       'ACER ASPIRE 3 RYZEN 3 DUAL CORE 3250U',
       'ASUS VIVOBOOK 14 (2022) CORE I5 12TH GEN',
       'ASUS VIVOBOOK 15 CORE I3 11TH GEN', 'ASUS CORE I7 11TH GEN',
       'DELL VOSTRO CORE I3 11TH GEN',
       'REALME BOOK (SLIM) CORE I3 11TH GEN',
       'HP 14S INTEL CORE I3 11TH GEN', 'HP RYZEN 5 HEXA CORE 5500U',
       'MSI BRAVO 15 RYZEN 5 HEXA CORE AMD R5',
       'ASUS ZENBOOK FLIP 14 OLED (2022) TOUCH PANEL CORE I5 12TH GEN',
       'HP 15S INTEL CORE I3 12TH GEN', 'HP INTEL CORE I5 11TH GEN',
       'HP RYZEN 3 QUAD CORE 5300U', 'LENOVO IDEAPAD 3 CORE I3 11TH GEN',
       'ASUS CORE I3 10TH GEN', 'HP ATHL

In [ ]:
df['ram'].unique()

array(['16 GB DDR4 RAM', '8 GB DDR4 RAM', '4 GB DDR4 RAM',
       '16 GB LPDDR5 RAM', '16 GB DDR5 RAM', '4 GB LPDDR4X RAM',
       '8 GB LPDDR4X RAM', '8 GB Unified Memory RAM', '32 GB LPDDR5 RAM',
       '32 GB DDR5 RAM', '4 GB LPDDR4 RAM', '16 GB LPDDR4X RAM',
       '8 GB DDR5 RAM', '32 GB Unified Memory RAM',
       '16 GB Unified Memory RAM', '16 GB LPDDR4 RAM', '32 GB DDR4 RAM',
       '8 GB DDR3 RAM', '16 GB LPDDR3 RAM', '8 GB LPDDR3 RAM',
       '32 GB LPDDR4X RAM', '16 GB DDR3 RAM'], dtype=object)

In [ ]:
df['os'].unique()

array(['WINDOWS 11 OPERATING SYSTEM',
       '64 BIT WINDOWS 11 OPERATING SYSTEM', 'DOS OPERATING SYSTEM',
       'MAC OS OPERATING SYSTEM', '64 BIT WINDOWS 10 OPERATING SYSTEM',
       '32 BIT WINDOWS 11 OPERATING SYSTEM', 'CHROME OPERATING SYSTEM',
       '64 BIT CHROME OPERATING SYSTEM', 'WINDOWS 10 OPERATING SYSTEM',
       '64 BIT DOS OPERATING SYSTEM', '64 BIT WINDOWS 8 OPERATING SYSTEM'],
      dtype=object)

In [ ]:
df['processor'].unique()

array(['INTEL CORE I5 PROCESSOR (11TH GEN)',
       'INTEL CORE I3 PROCESSOR (11TH GEN)',
       'INTEL CORE I5 PROCESSOR (10TH GEN)',
       'INTEL CORE I3 PROCESSOR (10TH GEN)',
       'AMD ATHLON DUAL CORE PROCESSOR', 'APPLE M1 PROCESSOR',
       'INTEL CELERON DUAL CORE PROCESSOR',
       'AMD RYZEN 3 QUAD CORE PROCESSOR',
       'AMD RYZEN 3 DUAL CORE PROCESSOR',
       'INTEL CORE I5 PROCESSOR (12TH GEN)',
       'INTEL CORE I7 PROCESSOR (11TH GEN)',
       'AMD RYZEN 5 HEXA CORE PROCESSOR',
       'INTEL CORE I3 PROCESSOR (12TH GEN)',
       'AMD RYZEN 7 OCTA CORE PROCESSOR',
       'QUALCOMM SNAPDRAGON 7C GEN 2 PROCESSOR',
       'INTEL CORE I7 PROCESSOR (12TH GEN)',
       'INTEL PENTIUM SILVER PROCESSOR',
       'AMD RYZEN 5 QUAD CORE PROCESSOR', 'APPLE M2 PROCESSOR',
       'INTEL CORE I9 PROCESSOR (12TH GEN)',
       'INTEL CORE I9 PROCESSOR (13TH GEN)',
       'INTEL CORE I7 PROCESSOR (13TH GEN)', 'AMD DUAL CORE PROCESSOR',
       'AMD RYZEN 9 OCTA CORE PROCESSOR', 'APPLE 

In [ ]:
df['storage'].unique()

array(['512 GB SSD', '1 TB HDD|256 GB SSD', '256 GB SSD', '1 TB SSD',
       '2 TB SSD', '1 TB HDD|512 GB SSD', '1 TB HDD', '128 GB SSD',
       '256 GB HDD|256 GB SSD', '4 TB SSD', '512 GB HDD|512 GB SSD',
       '1 TB HDD|128 GB SSD',
       'PCI-e SSD (NVMe) ready,Silver-Lining Print Keyboard,Matrix Display (Extend),Cooler Boost 5,Hi-Res Audio,Nahimic 3,144Hz Panel,Thin Bezel,RGB Gaming Keyboard,Speaker Tuning Engine,MSI Center',
       'PCI-e Gen4 SSD?SHIFT?Matrix Display (Extend)?Cooler Boost 3?Thunderbolt 4?Finger Print Security?True Color 2.0?Hi-Res Audio?Nahimic 3? 4-Sided Thin bezel?MSI Center?Silky Smooth Touchpad?Military-Grade Durability',
       '2 TB HDD', '256 GB HDD'], dtype=object)

In [ ]:
df1=df
df1.head()

name  price(in Rs.)  \
0             LENOVO INTEL CORE I5 11TH GEN          62990   
1            LENOVO V15 G2 CORE I3 11TH GEN          37500   
2      ASUS TUF GAMING F15 CORE I5 10TH GEN          49990   
3  ASUS VIVOBOOK 15 (2022) CORE I3 10TH GEN          33990   
4                   LENOVO ATHLON DUAL CORE          18990   

                            processor             ram  \
0  INTEL CORE I5 PROCESSOR (11TH GEN)  16 GB DDR4 RAM   
1  INTEL CORE I3 PROCESSOR (11TH GEN)   8 GB DDR4 RAM   
2  INTEL CORE I5 PROCESSOR (10TH GEN)   8 GB DDR4 RAM   
3  INTEL CORE I3 PROCESSOR (10TH GEN)   8 GB DDR4 RAM   
4      AMD ATHLON DUAL CORE PROCESSOR   4 GB DDR4 RAM   

                                   os              storage  display(in inch)  \
0         WINDOWS 11 OPERATING SYSTEM           512 GB SSD              15.6   
1  64 BIT WINDOWS 11 OPERATING SYSTEM  1 TB HDD|256 GB SSD              15.6   
2         WINDOWS 11 OPERATING SYSTEM           512 GB SSD              15.6   
3  64 BIT WINDOWS 11 OPERATING SYSTEM           512 GB SSD              15.6   
4                DOS OPERATING SYSTEM           256 GB SSD              14.0   

   rating  no_of_ratings  no_of_reviews  
0     4.5           14.0            1.0  
1     4.4           53.0            3.0  
2     4.4         4733.0          463.0  
3     4.3        10406.0         1040.0  
4     3.8           18.0            3.0

In [ ]:
temp=df['price(in Rs.)']
df=df.drop(columns='price(in Rs.)')

In [ ]:
label_encoder=preprocessing.LabelEncoder()
for i in df:
  df[i]=label_encoder.fit_transform(df[i])

In [ ]:
df.dtypes

name                int64
processor           int64
ram                 int64
os                  int64
storage             int64
display(in inch)    int64
rating              int64
no_of_ratings       int64
no_of_reviews       int64
dtype: object

In [ ]:
df.head()

name  processor  ram  os  storage  display(in inch)  rating  no_of_ratings  \
0   422         34    1  10       13                13      15             13   
1   438         30   17   4        2                13      14             49   
2   142         33   17  10       13                13      14            262   
3   162         29   17   4       13                13      13            270   
4   375          1   13   7       10                 6       7             17   

   no_of_reviews  
0              1  
1              3  
2            132  
3            144  
4              3

In [ ]:
scaler=StandardScaler()
for i in df:
  df[[i]]=scaler.fit_transform(df[[i]])
df.head()

name  processor       ram        os   storage  display(in inch)  \
0  1.112451   0.493488 -1.371270  1.860638  0.721124          0.529182   
1  1.226762   0.191972  0.825217 -0.444273 -1.792706          0.529182   
2 -0.887983   0.418109  0.825217  1.860638  0.721124          0.529182   
3 -0.745095   0.116593  0.825217 -0.444273  0.721124          0.529182   
4  0.776664  -1.994024  0.276096  0.708183  0.035534         -1.163807   

     rating  no_of_ratings  no_of_reviews  
0  0.902986      -1.339333      -1.079453  
1  0.562680      -0.909036      -1.029249  
2  0.562680       1.636887       2.208958  
3  0.222375       1.732508       2.510187  
4 -1.819459      -1.291522      -1.029249

In [ ]:
df['price(in Rs.)']=temp
df.head()

name  processor       ram        os   storage  display(in inch)  \
0  1.112451   0.493488 -1.371270  1.860638  0.721124          0.529182   
1  1.226762   0.191972  0.825217 -0.444273 -1.792706          0.529182   
2 -0.887983   0.418109  0.825217  1.860638  0.721124          0.529182   
3 -0.745095   0.116593  0.825217 -0.444273  0.721124          0.529182   
4  0.776664  -1.994024  0.276096  0.708183  0.035534         -1.163807   

     rating  no_of_ratings  no_of_reviews  price(in Rs.)  
0  0.902986      -1.339333      -1.079453          62990  
1  0.562680      -0.909036      -1.029249          37500  
2  0.562680       1.636887       2.208958          49990  
3  0.222375       1.732508       2.510187          33990  
4 -1.819459      -1.291522      -1.029249          18990

In [ ]:
x=df.drop(columns='price(in Rs.)')
x.head()

name  processor       ram        os   storage  display(in inch)  \
0  1.112451   0.493488 -1.371270  1.860638  0.721124          0.529182   
1  1.226762   0.191972  0.825217 -0.444273 -1.792706          0.529182   
2 -0.887983   0.418109  0.825217  1.860638  0.721124          0.529182   
3 -0.745095   0.116593  0.825217 -0.444273  0.721124          0.529182   
4  0.776664  -1.994024  0.276096  0.708183  0.035534         -1.163807   

     rating  no_of_ratings  no_of_reviews  
0  0.902986      -1.339333      -1.079453  
1  0.562680      -0.909036      -1.029249  
2  0.562680       1.636887       2.208958  
3  0.222375       1.732508       2.510187  
4 -1.819459      -1.291522      -1.029249

In [ ]:
y=df['price(in Rs.)']
y.head()

0    62990
1    37500
2    49990
3    33990
4    18990
Name: price(in Rs.), dtype: int64

In [ ]:
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=.25,random_state=7)

**Model Selection**

In [ ]:
LR_score=cross_val_score(LinearRegression(),X=x,y=y,cv=5)
DT_score=cross_val_score(DecisionTreeRegressor(),X=x,y=y,cv=5)
RF_score=cross_val_score(RandomForestRegressor(),X=x,y=y,cv=5)
KN_score=cross_val_score(KNeighborsRegressor(),X=x,y=y,cv=5)
AD_score=cross_val_score(AdaBoostRegressor(),X=x,y=y,cv=5)
GB_score=cross_val_score(GradientBoostingRegressor(),X=x,y=y,cv=5)
XG_score=cross_val_score(XGBRegressor(),X=x,y=y,cv=5)
svm_score=cross_val_score(SVR(),X=x,y=y,cv=5)

In [ ]:
print(f'Linear Regression : {np.round(np.mean(LR_score)*100,2)}%')
print(f'Decision Tree : {np.round(np.mean(DT_score)*100,2)}%')
print(f'Random Forest : {np.round(np.mean(RF_score)*100,2)}%')
print(f'KNN : {np.round(np.mean(KN_score)*100,2)}%')
print(f'Adaboost : {np.round(np.mean(AD_score)*100,2)}%')
print(f'Gradient Boost : {np.round(np.mean(GB_score)*100,2)}%')
print(f'XGB : {np.round(np.mean(XG_score)*100,2)}%')
print(f'SVM : {np.round(np.mean(svm_score)*100,2)}%')

Linear Regression : 26.24%
Decision Tree : 66.73%
Random Forest : 78.88%
KNN : 55.47%
Adaboost : 58.76%
Gradient Boost : 80.64%
XGB : 81.43%
SVM : -11.47%


We select 3 models:
* Random Forest : 78.88%
* Gradient Boost : 80.64%
* XGB : 81.43%

**Model Evaluation**

In [ ]:
rf=RandomForestRegressor()
forest_params=[{'max_depth':list(range(10,15)),'max_features':list(range(0,15))}]

In [ ]:
clf=GridSearchCV(rf,forest_params,cv=5,scoring='r2')
clf.fit(x_train,y_train)
print(clf.best_params_)
print(clf.best_score_)

/usr/local/lib/python3.12/dist-packages/sklearn/model_selection/_validation.py:528: FitFailedWarning: 
25 fits failed out of a total of 375.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
25 fits failed with the following error:
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/dist-packages/sklearn/model_selection/_validation.py", line 866, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "/usr/local/lib/python3.12/dist-packages/sklearn/base.py", line 1382, in wrapper
    estimator._validate_params()
  File "/usr/local/lib/python3.12/dist-packages/sklearn/base.py", line 436, in _validate_params
    validate_parameter_constraints(
  File "/usr/local/lib/python3.12/dist-packages/sklearn/utils/

{'max_depth': 12, 'max_features': 2}
0.8212935129263375


{'max_depth': 12, 'max_features': 2}
0.8212935129263375

In [ ]:
gb=GradientBoostingRegressor()
gb_params=[{'max_depth':list(range(2,7)),'learning_rate':[i/100 for i in range(1,21)]}] #(0.01-0.2)

In [ ]:
clf=GridSearchCV(gb,gb_params,cv=5,scoring='r2')
clf.fit(x_train,y_train)
print(clf.best_params_)
print(clf.best_score_)

{'learning_rate': 0.15, 'max_depth': 3}
0.8337438985744298


{'learning_rate': 0.15, 'max_depth': 3}
0.834320915972992

In [ ]:
xg=XGBRegressor()
xg_params=[{'colsample_bytree':[i/10 for i in range(4,11)],'gamma':list(range(0,10))}]

In [ ]:
clf=GridSearchCV(xg,xg_params,scoring='r2')
clf.fit(x_train,y_train)
print(clf.best_params_)
print(clf.best_score_)

{'colsample_bytree': 0.7, 'gamma': 0}
0.806666910648346


{'colsample_bytree': 0.7, 'gamma': 0}
0.806666910648346

After Applying Grid Search we have 2 model:
* Gradient Boost :{'learning_rate': 0.15, 'max_depth': 3}0.834320915972992
* Random Forest : {'max_depth': 14, 'max_features': 3}
0.8200453232509177

In [ ]:
GB=GradientBoostingRegressor(learning_rate=0.15,max_depth=3)
GB.fit(x_train,y_train)

GB_pred=GB.predict(x_test)
GB_pred

array([ 39942.98019411, 205653.21299568,  72231.18240376, 116097.27912499,
        60307.8375031 ,  76450.39758833,  33653.17398671,  76278.55460383,
        21588.38653812,  39292.57316307,  82906.90941187, 199637.39332278,
        41259.73753775,  28140.10465052,  41089.74674158,  68335.61851745,
       157594.54046172,  60307.8375031 ,  59433.70240077,  37626.30574271,
        67715.91507103,  57971.20897735,  53569.93243876,  35637.88180497,
        37807.33275822, 175366.90616437,  56058.3675862 ,  77387.49363579,
        60667.9442806 ,  43567.90204104,  70496.05401211,  96147.13864398,
        55347.16547426,  23065.29474929,  75387.05890562,  57847.73612861,
       132670.09570762,  38009.42598666,  54383.45594231, 124533.96762016,
        37468.09080917,  56352.02328927,  66286.92344772, 219926.97063466,
       124533.96762016,  60307.8375031 ,  60307.8375031 ,  60307.8375031 ,
        54347.3728061 ,  92990.56695612,  48609.33137558, 199637.39332278,
        80310.00507407,  

In [ ]:
a=GB.score(x_test,y_test)
print(f'Accuracy of the Gradient Boost Algorithm : {np.round((a*100),2)}%')

Accuracy of the Gradient Boost Algorithm : 86.54%


In [ ]:
RF=RandomForestRegressor(max_depth=14,max_features=3)
RF.fit(x_train,y_train)

RF_pred=RF.predict(x_test)
RF_pred

array([ 43404.4673131 , 196971.67857143,  61616.08933081, 127500.28      ,
        58847.10047791,  81372.98617231,  37854.43501216,  76349.38654335,
        23090.93096774,  41710.26684384,  76930.40605617, 189993.39666667,
        41118.99785351,  34885.30318613,  43574.85818761,  66544.22179024,
       147110.34      ,  58847.10047791,  56388.52384208,  35141.86413241,
        93338.25333333,  55882.83163804,  53145.5134689 ,  44515.619574  ,
        41199.12894749, 167804.6530303 ,  57509.15257143,  81898.43761722,
        63220.85295423,  44163.33111111,  62342.7941574 , 101621.25666667,
        51988.82697691,  36651.324     ,  69260.13376105,  58280.7146089 ,
       172671.71      ,  36879.02523591,  59833.13695346, 129104.01      ,
        43191.36269177,  54302.2256553 ,  70433.37333333, 204407.78333333,
       129104.01      ,  58847.10047791,  58847.10047791,  58847.10047791,
        62322.50728571,  94696.99559524,  54358.31649478, 189993.39666667,
        70600.00778768,  

In [ ]:
a=RF.score(x_test,y_test)
print(f'Random Forest Algorithm : {np.round((a*100),2)}%')

Random Forest Algorithm : 85.73%


**Model Deployment**

Gradient Boost is selected as the final Rgression Algorithm

In [ ]:
name=input("Enter the Name of the laptop along with specs, for eg:'Lenovo Intel Core i5 11th Gen':").upper()
processor=input('Enter the processor name:').upper()
ram=input('Enter the RAM:').upper()
os=input('Enter the operating system(OS):').upper()
storage=input('Enter the storage type:').upper()
display=float(input('Enter the display size (in inches):'))
rating=float(input('Enter the rating (0-5): '))
no_of_ratings=int(input('Enter the no_of_ratings:'))
no_of_reviews=int(input('Enter the no_of_reviews:'))

Enter the Name of the laptop along with specs, for eg:'Lenovo Intel Core i5 11th Gen':Lenovo Intel Core i5 11th Gen':ASUS TUF GAMING F15 CORE I5 10TH GEN
Enter the processor name:Intel Core i5 Processor (10th Gen)
Enter the RAM:16 GB LPDDR5 RAM
Enter the operating system(OS):64 BIT WINDOWS 11 OPERATING SYSTEM
Enter the storage type:512 GB HDD|512 GB SSD
Enter the display size (in inches):73.5
Enter the rating (0-5): 4.3
Enter the no_of_ratings:1000
Enter the no_of_reviews:1000


In [ ]:
final={
    "name":[name],
    "processor":[processor],
    "ram":[ram],
    "os":[os],
    "storage":[storage],
    "display(in inch)":[display],
    "rating":[rating],
    "no_of_ratings":[no_of_ratings],
    "no_of_reviews":[no_of_reviews]
}
data=pd.DataFrame(final,columns=x_train.columns)

In [ ]:
for i in data:
  data[i]=label_encoder.fit_transform(data[i])
for i in data:
  data[[i]]=scaler.fit_transform(data[[i]])

In [ ]:
p=GB.predict(data)
print(f'Price of the laptop is Rs {np.round(p)}')

Price of the laptop is Rs [151139.]
